### DISCUSSION TITLE
How to make your own theme?

### DISCUSSION THREAD

**Discussion by BytePetter** [AvaloniaUI/Avalonia#11973] [Q&A]
Hello,

the documents say nothing on how to go about creating your own theme for your app.
The only paragraph I found links to an old GitHub-Page that 404's. [https://docs.avaloniaui.net/docs/styling/resources#theme-resources](link)

So, how do you accomplish that?

**Comment by MikeCodesDotNET** [comment]
You'll want to look at the [existing themes](https://github.com/AvaloniaUI/Avalonia/tree/master/src/Avalonia.Themes.Fluent).

**Comment by BytePetter** [comment]
Thanks. Thats a little much to start with, tough. There's no simpler sample to make it easier to understand for beginners?

**Comment by davidbitterlich** [comment]
Hi, I was interested in the same topic. Seems like there is nothing - the only thing one could do is taking Fluent theme as an example and construct an own theme based on it... I would love to have some documentation, like a list of all style properties each control has or something like that. Or information in how to change shapes of controls, like rounder buttons (or more edgy)... With enough time it is possible to fully analyze Fluent theme and maybe read through the codebase to see which properties each control has, but that's a lot of effort...

A good documentation would be very beneficial :)

**Reply by maxkatz6** [reply]
This documentation isn't really possible.
You don't start writing theme from learning how to make button round. You start writing theme after you got hundreds of round buttons and want to extract these styles into some common theme.

There can be an example, which is planned, at best, but it also won't be able to cover everything. FluentTheme at this context is just an advanced example.

There are lots of documentations on building blocks already - ControlTheme, Styles, Resources.

**Reply by stevemonaco** [reply]
Available features can be documented, but it's up to the developer to know how to use them or figure it out. The more documentation, the more time is spent to maintain. I have some plans to write documentation on how to customize existing themes regarding what is and isn't possible, but that's a few large items down the list of side projects. I find information that broadly covers problem-solving approaches to be much more valuable than one that only covers a specific case or two.

**Comment by Shadowblitz16** [comment]
This is what is keeping me from using avalonia.
Avalonia's default themes don't cut it for me and there isn't any documentation on how to make my own theme.
There needs to be documentation on how to style each control and what their existing style properties are.

**Reply by rabbitism** [reply]
Avalonia built-in controls are lookless, there is no way to document how to style it, because you can redefine the styles entirely as long as you covers the required template parts. There are many 3rd party themes that don't rely on avalonia fluent/simple theme. You can check their source code as reference.

**Reply by MikeCodesDotNET** [reply]
There's plenty of docs on how to style controls, and lots of example themes. You should be able to work it out from there.

**Reply by Shadowblitz16** [reply]
> Avalonia built-in controls are lookless, there is no way to document how to style it, because you can redefine the styles entirely as long as you covers the required template parts. There are many 3rd party themes that don't rely on avalonia fluent/simple theme. You can check their source code as reference.

This doesn't make any sense since the examples use properties and controls that are undocumented.
If this is the case then there needs to be documentation on how the theme system works and how it relates to controls.

**Reply by Shadowblitz16** [reply]
> There's plenty of docs on how to style controls, and lots of example themes. You should be able to work it out from there.

Not really, I just looked at this https://docs.avaloniaui.net/docs/guides/custom-controls/how-to-create-templated-controls
And it doesn't really reveal much.

This is kindof the issue with dynamic weakly typed languages like axaml is new users can't tell what anything is.

**Reply by thevortexcloud** [reply]
> If this is the case then there needs to be documentation on how the theme system works and how it relates to controls.

I suggest reading WPF's docs on templates. It's a lot more detailed and works pretty closely to how it works in Avalonia.

https://learn.microsoft.com/en-us/dotnet/desktop/wpf/controls/how-to-create-apply-template?view=netframeworkdesktop-4.8
https://learn.microsoft.com/en-us/dotnet/api/system.windows.templatepartattribute?view=windowsdesktop-9.0

But the key thing you need is to define all the template parts of a control. You can find the template parts by looking at the C# code for the control. The attribute tells you what the type the part should be, if it's required, and what it must be called. EG here is the toggle switch parts:

https://github.com/AvaloniaUI/Avalonia/blob/c6cfcbacba78d9a515b41fb6b9d3434aa35aa34d/src/Avalonia.Controls/ToggleSwitch.cs#L11-L19

If you don't make a required part, the control likely wont work properly. Beyond that, you can do what ever you want with the control template.

The compiler should also complain if you are missing a template part in recent versions of Avalonia.
